File: src/main/java/tn/esprit/entities/reservation_historique.java
package tn.esprit.entities;

import java.util.Date;

/**
 * Entité représentant une entrée dans l'historique des modifications d'une réservation.
 * Correspond à la table `reservation_historique` en base de données.
 */
public class reservation_historique {

    private int    id_historique;
    private int    id_reservation;
    private String champ_modifie;
    private String ancienne_valeur;
    private String nouvelle_valeur;
    private String modifie_par;
    private Date   date_modification;
    private String commentaire;

    // ─── Constructeurs ─────────────────────────────────────────────────────────

    public reservation_historique() {}

    /** Constructeur principal utilisé lors de l'enregistrement d'une modification. */
    public reservation_historique(int idReservation, String champModifie,
                                  String ancienneValeur, String nouvelleValeur,
                                  String modifiePar) {
        this.id_reservation  = idReservation;
        this.champ_modifie   = champModifie;
        this.ancienne_valeur = ancienneValeur;
        this.nouvelle_valeur = nouvelleValeur;
        this.modifie_par     = modifiePar;
        this.date_modification = new Date();
    }

    // ─── Getters & Setters ─────────────────────────────────────────────────────

    public int getId_historique()               { return id_historique; }
    public void setId_historique(int v)         { this.id_historique = v; }

    public int getId_reservation()              { return id_reservation; }
    public void setId_reservation(int v)        { this.id_reservation = v; }

    public String getChamp_modifie()            { return champ_modifie; }
    public void setChamp_modifie(String v)      { this.champ_modifie = v; }

    public String getAncienne_valeur()          { return ancienne_valeur; }
    public void setAncienne_valeur(String v)    { this.ancienne_valeur = v; }

    public String getNouvelle_valeur()          { return nouvelle_valeur; }
    public void setNouvelle_valeur(String v)    { this.nouvelle_valeur = v; }

    public String getModifie_par()              { return modifie_par; }
    public void setModifie_par(String v)        { this.modifie_par = v; }

    public Date getDate_modification()          { return date_modification; }
    public void setDate_modification(Date v)    { this.date_modification = v; }

    public String getCommentaire()              { return commentaire; }
    public void setCommentaire(String v)        { this.commentaire = v; }

    @Override
    public String toString() {
        return "Historique{id=" + id_historique +
                ", reservation=" + id_reservation +
                ", champ='" + champ_modifie + "'" +
                ", " + ancienne_valeur + " → " + nouvelle_valeur +
                ", par='" + modifie_par + "'" +
                ", date=" + date_modification + "}";
    }
}

File: src/Entity/ReservationHistorique.php
<?php

namespace App\Entity;

use Doctrine\ORM\Mapping as ORM;

#[ORM\Entity]
#[ORM\Table(name: 'reservation_historique')]
#[ORM\HasLifecycleCallbacks]
class ReservationHistorique
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column(name: 'id_historique', type: 'integer')]
    /** @phpstan-ignore property.unusedType */
    private ?int $id = null;

    #[ORM\ManyToOne(targetEntity: ReservationLocal::class, inversedBy: 'historiques')]
    #[ORM\JoinColumn(name: 'reservation_id', referencedColumnName: 'id_reservation', nullable: false)]
    private ReservationLocal $reservation;

    #[ORM\Column(name: 'champ_modifie', type: 'string', length: 255)]
    private string $champModifie;

    #[ORM\Column(name: 'ancienne_valeur', type: 'text', nullable: true)]
    private ?string $ancienneValeur = null;

    #[ORM\Column(name: 'nouvelle_valeur', type: 'text', nullable: true)]
    private ?string $nouvelleValeur = null;

    #[ORM\Column(name: 'modifie_par', type: 'string', length: 255, nullable: true)]
    private ?string $modifiePar = null;

    #[ORM\Column(name: 'date_modification', type: 'datetime')]
    private \DateTimeInterface $dateModification;

    #[ORM\Column(type: 'string', length: 255, nullable: true)]
    private ?string $commentaire = null;

    public function getIdHistorique(): ?int { return $this->id; }
    public function getReservation(): ReservationLocal { return $this->reservation; }
    public function setReservation(ReservationLocal $v): static { $this->reservation = $v; return $this; }
    public function getChampModifie(): string { return $this->champModifie; }
    public function setChampModifie(string $v): static { $this->champModifie = $v; return $this; }
    public function getAncienneValeur(): ?string { return $this->ancienneValeur; }
    public function setAncienneValeur(?string $v): static { $this->ancienneValeur = $v; return $this; }
    public function getNouvelleValeur(): ?string { return $this->nouvelleValeur; }
    public function setNouvelleValeur(?string $v): static { $this->nouvelleValeur = $v; return $this; }
    public function getModifiePar(): ?string { return $this->modifiePar; }
    public function setModifiePar(?string $v): static { $this->modifiePar = $v; return $this; }
    public function getDateModification(): \DateTimeInterface { return $this->dateModification; }
    protected function setDateModification(\DateTimeInterface $v): static { $this->dateModification = $v; return $this; }

    #[ORM\PrePersist]
    public function initDateModification(): void
    {
        $this->dateModification = new \DateTime();
    }
    public function getCommentaire(): ?string { return $this->commentaire; }
    public function setCommentaire(?string $v): static { $this->commentaire = $v; return $this; }
}